File: leet_BFS_0103_zigzag_level_order.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root)
    {
        if (!root)
            return {};

        vector<vector<int>> r;
        deque<TreeNode *>   D{root};
        int                 count = 0;

        while (!D.empty())
        {
            vector<int> temp;
            int         len = D.size();
            while (len--)
            {
                TreeNode    *node = D.front();
                
                temp.push_back(node->val);
                D.pop_front();
                if (node->left)
                    D.push_back(node->left);
                if (node->right)
                    D.push_back(node->right);
            }
            if (count % 2)
                reverse(temp.begin(), temp.end());
            ++count;
            r.push_back(temp);

        }
        return r;
    }
};

File: leet_BFS_0103_zigzag_level_order.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        D = deque([root])
        count = 0
        r = []
        while D:
            temp = []
            size = len(D)
            while size > 0:
                node = D.popleft()
                temp.append(node.val)
                if node.left:
                    D.append(node.left)
                if node.right:
                    D.append(node.right)
                size -= 1
            if count % 2 ^ 0:
                r.append(temp[::-1])
            else:
                r.append(temp)
            count += 1
        return r
